pkg/ucp/store/postgres/postgresclient.go

```diff
@@ -0,0 +1,422 @@
+	}
+
+	var routingScopePrefixFilter *string
+	if query.RoutingScopePrefix != "" {
+		routingScopePrefixFilter = to.Ptr(storeutil.NormalizePart(query.RoutingScopePrefix))
+	}
+
+	var timestampFilter *string
+	if config.PaginationToken != "" {
+		ts, err := p.parsePaginationToken(config.PaginationToken)
+		if err != nil {
+			return nil, &store.ErrInvalid{Message: "invalid argument. 'query.PaginationToken' is invalid."}
+		}
+		timestampFilter = &ts
+	}
+
+	var limitFilter *int
+	if config.MaxQueryItemCount > 0 {
+		limitFilter = &config.MaxQueryItemCount
+	}
+
+	// For a scope query, we need to perform the same normalization as we do for other operations on scopes.
+	if query.IsScopeQuery {
+		var err error
+		query.ResourceType, err = storeutil.ConvertScopeTypeToResourceType(query.ResourceType)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// NOTE: building SQL by concatenating strings is hard to do safely and should be avoided.
+	// If you need to work on this code MAKE SURE you use SQL parameters
+	// for any user input.
+	sql := `
+SELECT original_id, etag, resource_data, created_at 
+FROM resources
+WHERE ((root_scope = $1) OR ($2 AND (root_scope LIKE $1 || '%'))) AND 
+	resource_type = $3 AND 
+	((routing_scope LIKE $4 || '%') OR $4 IS NULL) AND 
+	(created_at > $5::TIMESTAMP OR $5 IS NULL)
+ORDER BY created_at ASC
+LIMIT $6`
+
+	args := []any{
+		// If ScopeRecursive is false, the RootScope must match exactly.
+		// If ScopeRecursive is true, the RootScope must be a prefix of the stored RootScope.
+		storeutil.NormalizePart(query.RootScope),
+		query.ScopeRecursive,
+		resourceType,
+		routingScopePrefixFilter, // RoutingScopePrefix is optional and always treated as as prefix.
+		timestampFilter,          // Optional for pagination.
+		limitFilter,              // NOTE: Postgres allows LIMIT to be set with a NULL value to mean no limit.
+	}
+
+	rows, err := p.api.Query(ctx, sql, args...)
+	if err != nil {
+		return nil, err
+	}
+	defer rows.Close()
+
+	// Capture the last timestamp so we can use it for pagination.
+	var timestamp *time.Time
+
+	result := store.ObjectQueryResult{}
+	for rows.Next() {
+		obj := store.Object{}
+		err := rows.Scan(&obj.ID, &obj.ETag, &obj.Data, &timestamp)
+		if err != nil {
+			return nil, err
+		}
+
+		// We could improve this by moving the filter logic to the SQL query.
+		//
+		// The problem is that the current filter logic is not well documented or tested, and
+		// we want to stay compatible with the existing implementation for now.
+		match, err := obj.MatchesFilters(query.Filters)
+		if err != nil {
+			return nil, err
+		} else if !match {
+			continue
+		}
+
+		result.Items = append(result.Items, obj)
+	}
+
+	err = rows.Err()
+	if err != nil {
+		return nil, err
+	}
+
+	if len(result.Items) < config.MaxQueryItemCount && config.MaxQueryItemCount > 0 {
+		// No more rows, so no need for pagination.
+		return &result, nil
+	}
+
+	if timestamp != nil {
```

Comment:
UUID can also be good for internal operations. Like instead of using the timestamp for the pagination, we can use the UUID of the last row.

Reply:
Is this just proposing an alternative?